saleor/checkout/utils.py

```diff
@@ -1,27 +1,619 @@
+            initial={'country': billing_address.country},
+            instance=billing_address)
+        initial_address = AddressChoiceForm.NEW_ADDRESS
+    else:
+        address_form, preview = get_address_form(
+            data,
+            autocomplete_type='billing',
+            country_code=country.code,
+            initial={'country': country})
+        if cart.user and cart.user.default_billing_address:
+            initial_address = cart.user.default_billing_address.id
+        else:
+            initial_address = AddressChoiceForm.NEW_ADDRESS
+
+    addresses_form = AddressChoiceForm(
+        data, addresses=user_addresses, initial={'address': initial_address})
+    return address_form, addresses_form, preview
+
+
+def update_billing_address_in_cart(cart, user_addresses, data, country):
+    """Return shipping address choice forms and if an address was updated."""
+    address_form, addresses_form, preview = (
+        get_summary_without_shipping_forms(
+            cart, user_addresses, data, country))
+
+    updated = False
+
+    if addresses_form.is_valid():
+        use_existing_address = (
+            addresses_form.cleaned_data['address'] !=
+            AddressChoiceForm.NEW_ADDRESS)
+
+        if use_existing_address:
+            address_id = addresses_form.cleaned_data['address']
+            address = Address.objects.get(id=address_id)
+            change_billing_address_in_cart(cart, address)
+            updated = True
+
+        elif address_form.is_valid():
+            address = address_form.save()
+            change_billing_address_in_cart(cart, address)
+            updated = True
+
+    return addresses_form, address_form, updated
+
+
+def change_billing_address_in_cart(cart, address):
+    """Save billing address in cart if changed.
+
+    Remove previously saved address if not connected to any user.
+    """
+    has_address_changed = (
+        not address and cart.billing_address or
+        address and not cart.billing_address or
+        address and cart.billing_address and address != cart.billing_address)
+    if has_address_changed:
+        remove_old_address = (
+            cart.billing_address and (not cart.user or (
+                cart.user and
+                cart.billing_address not in cart.user.addresses.all())))
+        if remove_old_address:
+            cart.billing_address.delete()
+        cart.billing_address = address
+        cart.save(update_fields=['billing_address'])
+
+
+def change_shipping_address_in_cart(cart, address):
+    """Save shipping address in cart if changed.
+
+    Remove previously saved address if not connected to any user.
+    """
+    has_address_changed = (
+        not address and cart.shipping_address or
+        address and not cart.shipping_address or
+        address and cart.shipping_address and address != cart.shipping_address)
+    if has_address_changed:
+        remove_old_address = (
+            cart.shipping_address and (not cart.user or (
```

Comment:
Not a fan of such nesting, can we introduce some variables?

Reply:
I will extract a separate function checking if an address has changed and should be removed.